File: TDA-Estudiantes/CPP/Estudiante.cpp
#include "Estudiante.h"

#include <ctime>
#include <iostream>
#include <stdexcept>

using namespace std;

Estudiante::Estudiante() {
    cedula = "";
    nombres = "";
    apellidos = "";

    diaNacimiento = 1;
    mesNacimiento = 1;
    anioNacimiento = 1900;

    cantidadNotas = 0;

    for (int i = 0; i < MAX_NOTAS; i++) {
        notas[i] = 0.0;
    }
}

Estudiante::Estudiante(
    string cedula,
    string nombres,
    string apellidos,
    int diaNacimiento,
    int mesNacimiento,
    int anioNacimiento
) {
    if (cedula.empty()) {
        throw invalid_argument(
            "La cedula no puede estar vacia."
        );
    }

    if (nombres.empty()) {
        throw invalid_argument(
            "Los nombres no pueden estar vacios."
        );
    }

    if (apellidos.empty()) {
        throw invalid_argument(
            "Los apellidos no pueden estar vacios."
        );
    }

    if (!fechaValida(
            diaNacimiento,
            mesNacimiento,
            anioNacimiento)) {

        throw invalid_argument(
            "La fecha de nacimiento no es valida."
        );
    }

    this->cedula = cedula;
    this->nombres = nombres;
    this->apellidos = apellidos;

    this->diaNacimiento = diaNacimiento;
    this->mesNacimiento = mesNacimiento;
    this->anioNacimiento = anioNacimiento;

    cantidadNotas = 0;

    for (int i = 0; i < MAX_NOTAS; i++) {
        notas[i] = 0.0;
    }
}

bool Estudiante::fechaValida(
    int dia,
    int mes,
    int anio
) const {
    if (anio < 1900 || mes < 1 || mes > 12) {
        return false;
    }

    int diasPorMes[] = {
        31, 28, 31, 30, 31, 30,
        31, 31, 30, 31, 30, 31
    };

    bool bisiesto =
        (anio % 400 == 0)
        || (anio % 4 == 0 && anio % 100 != 0);

    if (bisiesto) {
        diasPorMes[1] = 29;
    }

    return dia >= 1
        && dia <= diasPorMes[mes - 1];
}

int Estudiante::calcularEdad() const {
    time_t tiempoActual = time(nullptr);
    tm* fechaActual = localtime(&tiempoActual);

    int anioActual = fechaActual->tm_year + 1900;
    int mesActual = fechaActual->tm_mon + 1;
    int diaActual = fechaActual->tm_mday;

    int edad = anioActual - anioNacimiento;

    bool aunNoCumple =
        mesActual < mesNacimiento
        || (mesActual == mesNacimiento
        && diaActual < diaNacimiento);

    if (aunNoCumple) {
        edad--;
    }

    return edad;
}

void Estudiante::mostrarDatos() const {
    cout << "Cedula: " << cedula << endl;
    cout << "Nombres: " << nombres << endl;
    cout << "Apellidos: " << apellidos << endl;

    cout << "Fecha de nacimiento: "
         << diaNacimiento << "/"
         << mesNacimiento << "/"
         << anioNacimiento << endl;

    cout << "Edad: "
         << calcularEdad() << endl;

    cout << "Cantidad de notas: "
         << cantidadNotas << "/"
         << MAX_NOTAS << endl;
}

string Estudiante::getCedula() const {
    return cedula;
}

string Estudiante::getNombres() const {
    return nombres;
}

string Estudiante::getApellidos() const {
    return apellidos;
}

int Estudiante::getDiaNacimiento() const {
    return diaNacimiento;
}

int Estudiante::getMesNacimiento() const {
    return mesNacimiento;
}

int Estudiante::getAnioNacimiento() const {
    return anioNacimiento;
}

int Estudiante::getCantidadNotas() const {
    return cantidadNotas;
}

int Estudiante::getCapacidadNotas() const {
    return MAX_NOTAS;
}

File: TDA-Estudiantes/Java/Estudiante.java
import java.time.DateTimeException;
import java.time.LocalDate;

public class Estudiante {

    public static final int MAX_NOTAS = 7;

    private String cedula;
    private String nombres;
    private String apellidos;

    private int diaNacimiento;
    private int mesNacimiento;
    private int anioNacimiento;

    // Vector estático de máximo siete calificaciones.
    private final double[] notas;
    private int cantidadNotas;

    public Estudiante(
            String cedula,
            String nombres,
            String apellidos,
            int diaNacimiento,
            int mesNacimiento,
            int anioNacimiento) {

        validarDatos(
                cedula,
                nombres,
                apellidos,
                diaNacimiento,
                mesNacimiento,
                anioNacimiento
        );

        this.cedula = cedula.trim();
        this.nombres = nombres.trim();
        this.apellidos = apellidos.trim();
        this.diaNacimiento = diaNacimiento;
        this.mesNacimiento = mesNacimiento;
        this.anioNacimiento = anioNacimiento;

        notas = new double[MAX_NOTAS];
        cantidadNotas = 0;
    }

    public void actualizarDatos(
            String cedula,
            String nombres,
            String apellidos,
            int diaNacimiento,
            int mesNacimiento,
            int anioNacimiento) {

        validarDatos(
                cedula,
                nombres,
                apellidos,
                diaNacimiento,
                mesNacimiento,
                anioNacimiento
        );

        this.cedula = cedula.trim();
        this.nombres = nombres.trim();
        this.apellidos = apellidos.trim();
        this.diaNacimiento = diaNacimiento;
        this.mesNacimiento = mesNacimiento;
        this.anioNacimiento = anioNacimiento;
    }

    private void validarDatos(
            String cedula,
            String nombres,
            String apellidos,
            int dia,
            int mes,
            int anio) {

        if (cedula == null || cedula.trim().isEmpty()) {
            throw new IllegalArgumentException(
                    "La cédula no puede estar vacía.");
        }

        if (nombres == null || nombres.trim().isEmpty()) {
            throw new IllegalArgumentException(
                    "Los nombres no pueden estar vacíos.");
        }

        if (apellidos == null || apellidos.trim().isEmpty()) {
            throw new IllegalArgumentException(
                    "Los apellidos no pueden estar vacíos.");
        }

        try {
            LocalDate fecha = LocalDate.of(anio, mes, dia);

            if (fecha.isAfter(LocalDate.now())) {
                throw new IllegalArgumentException(
                        "La fecha no puede ser futura.");
            }

        } catch (DateTimeException error) {
            throw new IllegalArgumentException(
                    "La fecha de nacimiento no es válida.");
        }
    }

    public int calcularEdad() {
        LocalDate nacimiento = LocalDate.of(
                anioNacimiento,
                mesNacimiento,
                diaNacimiento
        );

        LocalDate actual = LocalDate.now();

        int edad = actual.getYear() - nacimiento.getYear();

        boolean noHaCumplido =
                actual.getMonthValue()
                        < nacimiento.getMonthValue()
                || (actual.getMonthValue()
                        == nacimiento.getMonthValue()
                && actual.getDayOfMonth()
                        < nacimiento.getDayOfMonth());

        if (noHaCumplido) {
            edad--;
        }

        return edad;
    }

    public void mostrarDatos() {
        System.out.println("Cédula: " + cedula);
        System.out.println("Nombres: " + nombres);
        System.out.println("Apellidos: " + apellidos);

        System.out.println(
                "Fecha de nacimiento: "
                        + diaNacimiento + "/"
                        + mesNacimiento + "/"
                        + anioNacimiento);

        System.out.println("Edad: " + calcularEdad());

        System.out.println(
                "Notas registradas: "
                        + cantidadNotas + "/" + MAX_NOTAS);
    }

    public String getCedula() {
        return cedula;
    }

    public String getNombres() {
        return nombres;
    }

    public String getApellidos() {
        return apellidos;
    }

    public int getDiaNacimiento() {
        return diaNacimiento;
    }

    public int getMesNacimiento() {
        return mesNacimiento;
    }

    public int getAnioNacimiento() {
        return anioNacimiento;
    }

    public int getCantidadNotas() {
        return cantidadNotas;
    }
}